### ISSUE TITLE
Please backport fd1da784ac64 to 14.x ([compiler-rt] [scudo] Use -mcrc32 on x86 when available)

### ISSUE THREAD

**Issue by mgorny** [llvm/llvm-project#54719]
Please backport fd1da784ac644492f8ca40064baf3ef360352f55 to 14.x branch:

```
ommit fd1da784ac644492f8ca40064baf3ef360352f55 (HEAD -> main, origin/main)
Author: Michał Górny <mgorny@moritz.systems>
Date:   2022-03-31 09:55:25 +0200

    [compiler-rt] [scudo] Use -mcrc32 on x86 when available
    
    Update the hardware CRC32 logic in scudo to support using `-mcrc32`
    instead of `-msse4.2`.  The CRC32 intrinsics use the former flag
    in the newer compiler versions, e.g. in clang since 12fa608af44a.
    With these versions of clang, passing `-msse4.2` is insufficient
    to enable the instructions and causes build failures when `-march` does
    not enable CRC32 implicitly:
    
        /var/tmp/portage/sys-libs/compiler-rt-sanitizers-14.0.0/work/compiler-rt/lib/scudo/scudo_crc32.cpp:20:10: error: always_inline function '_mm_crc32_u32' requires target feature 'crc32', but would be inlined into function 'computeHardwareCRC32' that is compiled without support for 'crc32'
          return CRC32_INTRINSIC(Crc, Data);
                 ^
        /var/tmp/portage/sys-libs/compiler-rt-sanitizers-14.0.0/work/compiler-rt/lib/scudo/scudo_crc32.h:27:27: note: expanded from macro 'CRC32_INTRINSIC'
        #  define CRC32_INTRINSIC FIRST_32_SECOND_64(_mm_crc32_u32, _mm_crc32_u64)
                                  ^
        /var/tmp/portage/sys-libs/compiler-rt-sanitizers-14.0.0/work/compiler-rt/lib/scudo/../sanitizer_common/sanitizer_platform.h:132:36: note: expanded from macro 'FIRST_32_SECOND_64'
        #  define FIRST_32_SECOND_64(a, b) (a)
                                           ^
        1 error generated.
    
    For backwards compatibility, use `-mcrc32` when available and fall back
    to `-msse4.2`.  The `<smmintrin.h>` header remains in use as it still
    works and is compatible with GCC, while clang's `<crc32intrin.h>`
    is not.
    
    Use __builtin_ia32*() rather than _mm_crc32*() when using `-mcrc32`
    to preserve compatibility with GCC.  _mm_crc32*() are aliases
    to __builtin_ia32*() in both compilers but GCC requires `-msse4.2`
    for the former, while both use `-mcrc32` for the latter.
    
    Originally reported in https://bugs.gentoo.org/835870.
    
    Differential Revision: https://reviews.llvm.org/D122789
```
```

**Comment by tstellar** [comment]
@phoebewang Is this safe to backport? https://github.com/llvm/llvm-project/commit/fd1da784ac644492f8ca40064baf3ef360352f55

**Comment by phoebewang** [comment]
I think it is safe to backport. Thanks!

**Comment by tstellar** [comment]
Merged: 6697c5bc3a1e
